File: list_of_depths.go
package main

// Create an unbalanced tree:
//     1            // level = 0  [1]
//    / \
//   2   3         // level = 1 [2,3]
//  /
// 4              // level =2 [4]

// BFS

func listOfDepths(t *TreeNode) []*ListNode {
	if t == nil {
		return nil
	}
	first := &ListNode{
		Val: t.Val.(int),
	}
	mapOutput := map[int]*ListNode{
		0: first,
	}
	queue := []levelQueue{{Node: t.Left, Level: 1}, {Node: t.Right, Level: 1}}
	// queue       						mapOutput  		peak
	//{2.1,3.1}							{0:1}			2.1
	// {3.1, 4.2, nil.2}				{0:1, 1:2}
	// { 4.2, nil.2}					{0:1, 1:2-3}    3.1
	// {nil.2}											4.2
	for len(queue) > 0 {
		peak := queue[0]
		queue = queue[1:]
		if peak.Node == nil {
			continue
		}

		val, ok := mapOutput[peak.Level]
		node := &ListNode{
			Val: peak.Node.Val.(int),
		}
		if !ok {
			mapOutput[peak.Level] = node
		} else {
			head := val
			// move until the last element
			for val.Next != nil {
				val = val.Next
			}

			val.Next = node

			mapOutput[peak.Level] = head
		}

		if peak.Node.Right != nil || peak.Node.Left != nil {
			queue = append(queue, levelQueue{Node: peak.Node.Left, Level: peak.Level + 1}, levelQueue{Node: peak.Node.Right, Level: peak.Level + 1})
		}
	}

	output := []*ListNode{}
	for _, val := range mapOutput {
		output = append(output, val)
	}

	return output
}

type levelQueue struct {
	Node  *TreeNode
	Level int
}

File: leetcode/list_of_depths.py
import unittest
from typing import List, Optional
from inorder_traversal import TreeNode
from add_two_numbers import ListNode

class Solution:
    def listOfDepths(self, tree:Optional[TreeNode])->List[ListNode]:
        if not tree:
            return None
        return self.dfs(tree, [], 0)
    
    
    def dfs(self, node: Optional[TreeNode], lists:List[ListNode], level:int )->List[ListNode]:
        if not node:
            return lists
        
        if level < len(lists):
            curr = lists[level]
            p = curr
            while p:
                if not p.next:
                    p.next = ListNode(node.val)
                    break
                else:
                    p = p.next
        else:
            lists.append(ListNode(node.val))

        
        lists = self.dfs(node.left, lists, level + 1)
        return self.dfs(node.right, lists, level + 1)
            

class TestListOfDepths(unittest.TestCase):
    def setUp(self) -> None:
        self.sln = Solution()

    def linked_list_to_list(self, head: Optional[ListNode]):
        result = []
        node = head
        while node:
            result.append(node.val)
            node = node.next
        return result

    def levels_to_lists(self, levels):
        return [self.linked_list_to_list(head) if head else [] for head in (levels or [])]

    def test_empty_tree(self):
        root = None
        result = self.sln.listOfDepths(root)
        expected = []
        if result is None:
            self.assertEqual([], expected)
        else:
            self.assertEqual(self.levels_to_lists(result), expected)

    def test_single_node(self):
        root = TreeNode(1)
        result = self.sln.listOfDepths(root)
        self.assertIsNotNone(result)
        self.assertEqual(self.levels_to_lists(result), [[1]])

    def test_complete_three_levels(self):
        root = TreeNode(1)
        root.left = TreeNode(2)
        root.right = TreeNode(3)
        root.left.left = TreeNode(4)
        root.left.right = TreeNode(5)
        root.right.left = TreeNode(6)
        root.right.right = TreeNode(7)
        result = self.sln.listOfDepths(root)
        self.assertIsNotNone(result)
        self.assertEqual(self.levels_to_lists(result), [[1], [2, 3], [4, 5, 6, 7]])

    def test_left_skewed(self):
        root = TreeNode(1)
        root.left = TreeNode(2)
        root.left.left = TreeNode(3)
        root.left.left.left = TreeNode(4)
        result = self.sln.listOfDepths(root)
        self.assertIsNotNone(result)
        self.assertEqual(self.levels_to_lists(result), [[1], [2], [3], [4]])

    def test_right_skewed(self):
        root = TreeNode(1)
        root.right = TreeNode(2)
        root.right.right = TreeNode(3)
        root.right.right.right = TreeNode(4)
        result = self.sln.listOfDepths(root)
        self.assertIsNotNone(result)
        self.assertEqual(self.levels_to_lists(result), [[1], [2], [3], [4]])

    def test_sparse_tree(self):
        root = TreeNode(1)
        root.left = TreeNode(2)
        root.right = TreeNode(3)
        root.left.right = TreeNode(5)
        result = self.sln.listOfDepths(root)
        self.assertIsNotNone(result)
        self.assertEqual(self.levels_to_lists(result), [[1], [2, 3], [5]])


if __name__ == '__main__':
    unittest.main()
